DisplayList replay does not apply filters

nikolaszimmermann opened the pull request:
#### 21530f9cc92c09f19c293d2943fea77f31358b98
<pre>
DisplayList replay does not apply filters
<a href="https://bugs.webkit.org/show_bug.cgi?id=269738">https://bugs.webkit.org/show_bug.cgi?id=269738</a>

Reviewed by Said Abou-Hallawa.

DisplayList recording properly handles filters -- they
are serialized as &apos;DrawFilteredImageBuffer&apos; command. During replay this
display list item type is unhandled, therefore the filtered image is not
drawn when replaying the display list - fix that.

This blocks the intrduction of the multi-threaded CPU rendering mode
for WPE/Skia, since that relies on DisplayList recording &amp; replay.

Coverered by existing tests, when enforcing DisplayList recording.

* Source/WebCore/platform/graphics/displaylists/DisplayListItem.cpp:
(WebCore::DisplayList::applyFilteredImageBufferItem):
(WebCore::DisplayList::applyItem):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::DrawFilteredImageBuffer::apply const):
(WebCore::DisplayList::DrawFilteredImageBuffer::apply): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:

Canonical link: <a href="https://commits.webkit.org/275163@main">https://commits.webkit.org/275163@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/685655389f4d4b1c9ec640536a3854dad7964919

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40624 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19637 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43002 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43177 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36713 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42931 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22600 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16968 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33703 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41198 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16573 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35018 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14290 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14366 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35993 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44452 "Built successfully") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/34077 "Build is in progress. Recent messages:") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36836 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36319 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40058 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15439 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12658 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38391 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17058 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9219 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17109 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16702 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/685655389f4d4b1c9ec640536a3854dad7964919. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/24765

nikolaszimmermann commented:
@nikolaszimmermann I'm using DisplayList recording/replay facilities to implement multi-threaded CPU rendering in WPE/Skia - mostly testing MotionMark - which works flawlessly, except for the "developer.html" front page - when hovering over the "Run benchmark" button it disappears. The reason is that the button hover pseudo style activates a CSS filter on the button (brightness(..%) IIRC) - that triggers the use of `GraphicsContext::drawFilteredImageBuffer` which leads to the disappearing button since filteres are a no-op during replay (if the filter operation was recorded, and not the _result_ of the filter operation - which is used if e.g. feImage is used).

nikolaszimmermann commented:
I also tested on macOS using MiniBrowser, and activated "Use GPU process for DOM rendering", but that had no problem with the MotionMark page -- that is puzzling me, but I didn't investigate further. I was under the impression the aforementioned option implies DisplayList recording/replay for GraphicsLayerCA...

Hope you can enlighten me @shallawa  :-)

shallawa commented on Source/WebCore/platform/graphics/displaylists/DisplayListItem.cpp:
```diff
@@ -50,6 +51,19 @@ bool isValid(const Item& item)
+    RELEASE_ASSERT(item.sourceImageIdentifier().has_value());
+    auto resourceIdentifier = item.sourceImageIdentifier().value();
+    if (auto* sourceImage = resourceHeap.getImageBuffer(resourceIdentifier)) {
+        FilterResults results;
+        item.apply(context, sourceImage, results);
```
I think allocating new `FilterResults` for every `Filter::apply()` is expensive if it is an `SVGFilter`. Should we cache the results like what we do in `RemoteDisplayListRecorder::drawFilteredImageBuffer()`?

nikolaszimmermann replied:
Ah, good suggestion. I was already wdnering about that, but didn't follow it - as I have other tasks pending :-)
Filed https://bugs.webkit.org/show_bug.cgi?id=269905, to take care of that. Would like to land this now, to unblock further Skia/CPU+WPE work.

shallawa commented:
> I also tested on macOS using MiniBrowser, and activated "Use GPU process for DOM rendering", but that had no problem with the MotionMark page -- that is puzzling me, but I didn't investigate further. I was under the impression the aforementioned option implies DisplayList recording/replay for GraphicsLayerCA...
> 
> Hope you can enlighten me @shallawa :-)

Yes because we replay the display list items for GPUP through IPC messages. `RemoteDisplayListRecorderProxy::recordDrawFilteredImageBuffer()` sends the message `Messages::RemoteDisplayListRecorder::DrawFilteredImageBuffer`. When it is received, the sourceImages of all FEImage and the target sourceImage have to be resolved first before calling `DrawFilteredImageBuffer::apply()`.

Thanks for taking care of this. And it is good to know that display lists can be fully used outside GPUP.

nikolaszimmermann committed:
DisplayList replay does not apply filters
https://bugs.webkit.org/show_bug.cgi?id=269738

Reviewed by Said Abou-Hallawa.

DisplayList recording properly handles filters -- they
are serialized as 'DrawFilteredImageBuffer' command. During replay this
display list item type is unhandled, therefore the filtered image is not
drawn when replaying the display list - fix that.

This blocks the intrduction of the multi-threaded CPU rendering mode
for WPE/Skia, since that relies on DisplayList recording & replay.

Coverered by existing tests, when enforcing DisplayList recording.

* Source/WebCore/platform/graphics/displaylists/DisplayListItem.cpp:
(WebCore::DisplayList::applyFilteredImageBufferItem):
(WebCore::DisplayList::applyItem):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::DrawFilteredImageBuffer::apply const):
(WebCore::DisplayList::DrawFilteredImageBuffer::apply): Deleted.
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:

Canonical link: https://commits.webkit.org/275163@main

webkit-commit-queue commented:
Committed 275163@main (21530f9cc92c): <https://commits.webkit.org/275163@main>

Reviewed commits have been landed. Closing PR #24765 and removing active labels.
